TypeScript files are seen as JavaScript when tsup resolves types of an external package in certain conditions

If a package points a subpath export (`exports: { "./a": "./a.ts" }` in package.json) to a TypeScript file (“a.ts”: `import {…} from "./b.ts"`), and you use that package as a dependency, when you run tsup with `dts: { resolve: true }`:
the file (“a.ts”) itself is treated as TypeScript code, but TypeScript files (like “b.ts”) imported by the former file are treated as plain JavaScript codes.

If “a.ts” is set as the main export (`exports: { ".": "./ts" }`), everything works.

## Reproduction

https://github.com/umajho/tsup-reproduction-dts-resolve-pnpm-workspace

in “packages/support-lib”:
``` ts
// foo.ts:
export type Foo = "foo";
export const foo: Foo = "foo";

// bar-that-imports-foo.ts:
export * from "./foo";

// package.json:
{
  // …
  "exports": {
    "./foo": "./foo.ts",
    "./bar-that-imports-foo": "./bar-that-imports-foo.ts",
    ".": "./bar-that-imports-foo.ts"
  }
}
```

---

`cd packages/lib-to-be-built` 

### `pnpm run build-foo-dts-resolve` ⭕️

Importing from “support-lib” with the subpath “./foo”, and building the library using tsup with `dts: { resolve: true }`:

``` ts
// lib-imports-foo.ts:
import { Foo, foo } from "@tsup-reproduction-dts-resolve-pnpm-workspace/support-lib/foo";
export const baz: Foo = foo;
```

<details>
<summary>success</summary>

```
➜  lib-to-be-built git:(main) pnpm run build-foo-dts-resolve

> @tsup-reproduction-dts-resolve-pnpm-workspace/lib-to-be-built@1.0.0 build-foo-dts-resolve /private/tmp/test/packages/lib-to-be-built
> tsup --config tsup.foo-dts-resolve.config.ts

CLI Building entry: lib-imports-foo.ts
CLI Using tsconfig: tsconfig.json
CLI tsup v7.2.0
CLI Using tsup config: /private/tmp/test/packages/lib-to-be-built/tsup.foo-dts-resolve.config.ts
CLI Target: es2020
CLI Cleaning output folder
ESM Build start
ESM dist/lib-imports-foo.js 98.00 B
ESM ⚡️ Build success in 10ms
DTS Build start
DTS ⚡️ Build success in 474ms
DTS dist/lib-imports-foo.d.ts 60.00 B
```
</details>

### `pnpm run build-bar-using-main-dts-resolve` ⭕️

Importing from “support-lib” with the main path, and building the library using tsup with `dts: { resolve: true }`:

``` ts
// lib-imports-bar-using-main.ts:
import { Foo, foo } from "@tsup-reproduction-dts-resolve-pnpm-workspace/support-lib";
export const baz: Foo = foo;
```

<details>
<summary>success</summary>

```
➜  lib-to-be-built git:(main) pnpm run build-bar-using-main-dts-resolve

> @tsup-reproduction-dts-resolve-pnpm-workspace/lib-to-be-built@1.0.0 build-bar-using-main-dts-resolve /private/tmp/test/packages/lib-to-be-built
> tsup --config tsup.bar-using-main-dts-resolve.config.ts

CLI Building entry: lib-imports-bar-using-main.ts
CLI Using tsconfig: tsconfig.json
CLI tsup v7.2.0
CLI Using tsup config: /private/tmp/test/packages/lib-to-be-built/tsup.bar-using-main-dts-resolve.config.ts
CLI Target: es2020
CLI Cleaning output folder
ESM Build start
ESM dist/lib-imports-bar-using-main.js 109.00 B
ESM ⚡️ Build success in 12ms
DTS Build start
DTS ⚡️ Build success in 453ms
DTS dist/lib-imports-bar-using-main.d.ts 60.00 B
```

</details>

### `pnpm run build-bar-dts-resolve` ❌

Importing from “support-lib” with the subpath “./bar-that-imports-foo”, and building the library using tsup with `dts: { resolve: true }`:

``` ts
// lib-imports-bar.ts:
import { Foo, foo } from "@tsup-reproduction-dts-resolve-pnpm-workspace/support-lib/bar-that-imports-foo";
export const baz: Foo = foo;
```

<details open>
<summary>failed</summary>

```
➜  lib-to-be-built git:(main) pnpm run build-bar-dts-resolve

> @tsup-reproduction-dts-resolve-pnpm-workspace/lib-to-be-built@1.0.0 build-bar-dts-resolve /private/tmp/test/packages/lib-to-be-built
> tsup --config tsup.bar-dts-resolve.config.ts

CLI Building entry: lib-imports-bar.ts
CLI Using tsconfig: tsconfig.json
CLI tsup v7.2.0
CLI Using tsup config: /private/tmp/test/packages/lib-to-be-built/tsup.bar-dts-resolve.config.ts
CLI Target: es2020
CLI Cleaning output folder
ESM Build start
ESM dist/lib-imports-bar.js 98.00 B
ESM ⚡️ Build success in 16ms
DTS Build start
Error parsing: /private/tmp/test/packages/support-lib/foo.ts:1:7
Error: error occured in dts build
    at Worker.<anonymous> (/private/tmp/test/node_modules/.pnpm/tsup@7.2.0_typescript@5.2.2/node_modules/tsup/dist/index.js:2294:26)
    at Worker.emit (node:events:511:28)
    at MessagePort.<anonymous> (node:internal/worker:263:53)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:762:20)
    at exports.emitMessage (node:internal/per_context/messageport:23:28)
DTS Build error
Unexpected token (Note that you need plugins to import files that are not JavaScript)
1: export type Foo = "foo";
          ^
2: 
3: export const foo: Foo = "foo";
 ELIFECYCLE  Command failed with exit code 1.
```

</details>

<!-- POLAR PLEDGE BADGE START -->
## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/egoist) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/egoist/tsup/issues/1023">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/egoist/tsup/issues/1023/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/egoist/tsup/issues/1023/pledge.svg">
</picture>
</a>
<!-- POLAR PLEDGE BADGE END -->

I just noticed that when I set
```json
    "module": "NodeNext",
    "moduleResolution": "NodeNext",
```
in tsconfig.json, it just make the command not complain, but the generated `.d.ts` is wrong.

`pnpm run build-bar-dts-resolve`

generated definition file:
```js
import { Foo } from './foo.js'; // Cannot find module './foo.js' or its corresponding type declarations.ts(2307)

declare const baz: Foo;

export { baz };

```

--- a/src/api-extractor.ts
+++ b/src/api-extractor.ts
@@ -88,10 +88,18 @@ async function rollupDtsFiles(
   exports: ExportDeclaration[],
   format: Format,
 ) {
+  if (!options.experimentalDts || !options.experimentalDts?.entry) {
+    return
+  }
+
+  /**
+   * **`.tsup/declaration`** directory
+   */
   const declarationDir = ensureTempDeclarationDir()
   const outDir = options.outDir || 'dist'
   const pkg = await loadPkg(process.cwd())
   const dtsExtension = defaultOutExtension({ format, pkgType: pkg.type }).dts
+  const tsconfig = options.tsconfig || 'tsconfig.json'
 
   let dtsInputFilePath = path.join(
     declarationDir,
@@ -113,16 +121,40 @@ async function rollupDtsFiles(
     formatAggregationExports(exports, declarationDir),
   )
 
-  rollupDtsFile(
-    dtsInputFilePath,
-    dtsOutputFilePath,
-    options.tsconfig || 'tsconfig.json',
-  )
+  rollupDtsFile(dtsInputFilePath, dtsOutputFilePath, tsconfig)
 
   for (let [out, sourceFileName] of Object.entries(
-    options.experimentalDts!.entry,
+    options.experimentalDts.entry,
   )) {
+    /**
+     * **Source file name** (`src/index.ts`)
+     *
+     * @example
+     *
+     * ```ts
+     * import { defineConfig } from 'tsup'
+     *
+     * export default defineConfig({
+     *   entry: { index: 'src/index.ts' },
+     *   // Here `src/index.ts` is our `sourceFileName`.
+     * })
+     * ```
+     */
     sourceFileName = toAbsolutePath(sourceFileName)
+    /**
+     * **Output file name** (`dist/index.d.ts`)
+     *
+     * @example
+     *
+     * ```ts
+     * import { defineConfig } from 'tsup'
+     *
+     * export default defineConfig({
+     *  entry: { index: 'src/index.ts' },
+     * // Here `dist/index.d.ts` is our `outFileName`.
+     * })
+     * ```
+     */
     const outFileName = path.join(outDir, out + dtsExtension)
 
     // Find all declarations that are exported from the current source file

--- a/src/api-extractor.ts
+++ b/src/api-extractor.ts
@@ -93,7 +93,7 @@ async function rollupDtsFiles(
   }
 
   /**
-   * **`.tsup/declaration`** directory
+   * `.tsup/declaration` directory
    */
   const declarationDir = ensureTempDeclarationDir()
   const outDir = options.outDir || 'dist'
@@ -127,7 +127,7 @@ async function rollupDtsFiles(
     options.experimentalDts.entry,
   )) {
     /**
-     * **Source file name** (`src/index.ts`)
+     * Source file name (`src/index.ts`)
      *
      * @example
      *
@@ -142,7 +142,7 @@ async function rollupDtsFiles(
      */
     sourceFileName = toAbsolutePath(sourceFileName)
     /**
-     * **Output file name** (`dist/index.d.ts`)
+     * Output file name (`dist/index.d.ts`)
      *
      * @example
      *

--- a/src/exports.ts
+++ b/src/exports.ts
@@ -1,5 +1,5 @@
 import path from 'node:path'
-import { slash, trimDtsExtension, truthy } from './utils'
+import { replaceDtsWithJsExtensions, slash, truthy } from './utils'
 
 export type ExportDeclaration = ModuleExport | NamedExport
 
@@ -41,7 +41,7 @@ function formatAggregationExport(
   declaration: ExportDeclaration,
   declarationDirPath: string,
 ): string {
-  const dest = trimDtsExtension(
+  const dest = replaceDtsWithJsExtensions(
     `./${path.posix.normalize(
       slash(path.relative(declarationDirPath, declaration.destFileName)),
     )}`,
@@ -72,7 +72,7 @@ export function formatDistributionExports(
   fromFilePath: string,
   toFilePath: string,
 ) {
-  let importPath = trimDtsExtension(
+  let importPath = replaceDtsWithJsExtensions(
     path.posix.relative(
       path.posix.dirname(path.posix.normalize(slash(fromFilePath))),
       path.posix.normalize(slash(toFilePath)),

--- a/src/exports.ts
+++ b/src/exports.ts
@@ -48,7 +48,7 @@ function formatAggregationExport(
   )
 
   if (declaration.kind === 'module') {
-    // No implemeted
+    // Not implemented
     return ''
   } else if (declaration.kind === 'named') {
     return [

--- a/src/index.ts
+++ b/src/index.ts
@@ -12,8 +12,9 @@ import {
   type MaybePromise,
   debouncePromise,
   removeFiles,
+  resolveExperimentalDtsConfig,
+  resolveInitialExperimentalDtsConfig,
   slash,
-  toObjectEntry,
 } from './utils'
 import { createLogger, setSilent } from './log'
 import { runEsbuild } from './esbuild'
@@ -92,20 +93,10 @@ const normalizeOptions = async (
         : typeof _options.dts === 'string'
           ? { entry: _options.dts }
           : _options.dts,
-    experimentalDts: _options.experimentalDts
-      ? typeof _options.experimentalDts === 'boolean'
-        ? _options.experimentalDts
-          ? { entry: {} }
-          : undefined
-        : typeof _options.experimentalDts === 'string'
-          ? {
-              entry: toObjectEntry(_options.experimentalDts),
-            }
-          : {
-              ..._options.experimentalDts,
-              entry: toObjectEntry(_options.experimentalDts.entry || {}),
-            }
-      : undefined,
+
+    experimentalDts: await resolveInitialExperimentalDtsConfig(
+      _options.experimentalDts,
+    ),
   }
 
   setSilent(options.silent)
@@ -151,17 +142,14 @@ const normalizeOptions = async (
         ...(options.dts.compilerOptions || {}),
       }
     }
+
     if (options.experimentalDts) {
-      options.experimentalDts.compilerOptions = {
-        ...(tsconfig.data.compilerOptions || {}),
-        ...(options.experimentalDts.compilerOptions || {}),
-      }
-      options.experimentalDts.entry = toObjectEntry(
-        Object.keys(options.experimentalDts.entry).length > 0
-          ? options.experimentalDts.entry
-          : options.entry,
+      options.experimentalDts = await resolveExperimentalDtsConfig(
+        options as NormalizedOptions,
+        tsconfig,
       )
     }
+
     if (!options.target) {
       options.target = tsconfig.data?.compilerOptions?.target?.toLowerCase()
     }

--- a/src/index.ts
+++ b/src/index.ts
@@ -252,7 +252,7 @@ export async function build(_options: Options) {
               worker.on('message', (data) => {
                 if (data === 'error') {
                   terminateWorker()
-                  reject(new Error('error occured in dts build'))
+                  reject(new Error('error occurred in dts build'))
                 } else if (data === 'success') {
                   terminateWorker()
                   resolve()

--- a/src/tsc.ts
+++ b/src/tsc.ts
@@ -174,6 +174,7 @@ function emit(compilerOptions?: any, tsconfig?: string) {
       ...rawTsconfig.data,
       compilerOptions: {
         ...rawTsconfig.data?.compilerOptions,
+        ...compilerOptions,
 
         // Enable declaration emit and disable javascript emit
         noEmit: false,

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -279,7 +279,7 @@ export function replaceDtsWithJsExtensions(dtsFilePath: string) {
 }
 
 /**
- * Converts an array of {@linkcode NormalizedOptions.entry | entry paths}
+ * Converts an array of {@link NormalizedOptions.entry | entry paths}
  * into an object where the keys represent the output
  * file names (without extensions) and the values
  * represent the corresponding input file paths.
@@ -288,14 +288,13 @@ export function replaceDtsWithJsExtensions(dtsFilePath: string) {
  * @returns An object where the keys are the output file name and the values are the input file name.
  *
  * @example
- * <caption>### Convert an array of entry points to an entry point object</caption>
  *
  * ```ts
  * import { defineConfig } from 'tsup'
  *
  * export default defineConfig({
- *   entry: ['src/index.ts', 'src/types.ts'], // Becomes `{ index: 'src/index.ts', types: 'src/types.ts' }`
- *   format: ['esm', 'cjs'],
+ *   entry: ['src/index.ts', 'src/types.ts'],
+ *   // Becomes `{ index: 'src/index.ts', types: 'src/types.ts' }`
  * })
  * ```
  *
@@ -321,15 +320,10 @@ const convertArrayEntriesToObjectEntries = (arrayOfEntries: string[]) => {
 }
 
 /**
- * Resolves and standardizes entry paths into an object format.
- * If the provided entry is a string or an array of strings, it resolves
- * any potential glob patterns using
- * {@linkcode glob | tiny-glob's glob function}
- * and converts the result into an entry object.
- * If the input is already an object, it is returned as-is.
- *
- * @param entryPaths - The entry paths to resolve. Can be a string, an array of strings, or an object.
- * @returns A {@linkcode Promise} that resolves to the standardized entry paths in object format.
+ * Resolves and standardizes entry paths into an object format. If the provided
+ * entry is a string or an array of strings, it resolves any potential glob
+ * patterns amd converts the result into an entry object. If the input is
+ * already an object, it is returned as-is.
  *
  * @example
  *
@@ -340,7 +334,6 @@ const convertArrayEntriesToObjectEntries = (arrayOfEntries: string[]) => {
  *   entry: { index: 'src/index.ts' },
  *   format: ['esm', 'cjs'],
  *   experimentalDts: { entry: 'src/**\/*.ts' },
- *   // experimentalDts: { entry: 'src/**\/*.ts' }
  *   // becomes experimentalDts: { entry: { index: 'src/index.ts', types: 'src/types.ts } }
  * })
  * ```
@@ -358,20 +351,20 @@ const resolveEntryPaths = async (entryPaths: InputOption) => {
 
 /**
  * Resolves the
- * {@linkcode NormalizedExperimentalDtsConfig | experimental DTS config}
- * by resolving entry paths and merging the provided
- * TypeScript configuration options.
+ * {@link NormalizedExperimentalDtsConfig | experimental DTS config} by
+ * resolving entry paths and merging the provided TypeScript configuration
+ * options.
  *
- * @param options - The options containing entry points and experimental DTS configuration.
+ * @param options - The options containing entry points and experimental DTS
+ * configuration.
  * @param tsconfig - The loaded TypeScript configuration data.
- * @returns A {@linkcode Promise | promise} that resolves to the normalized experimental DTS configuration, or `undefined` if no entry or experimental DTS option is provided.
  *
  * @internal
  */
 export const resolveExperimentalDtsConfig = async (
   options: NormalizedOptions,
   tsconfig: any,
-): Promise<NormalizedOptions['experimentalDts']> => {
+): Promise<NormalizedExperimentalDtsConfig> => {
   const resolvedEntryPaths = await resolveEntryPaths(
     options.experimentalDts?.entry || options.entry,
   )
@@ -397,29 +390,14 @@ export const resolveExperimentalDtsConfig = async (
 }
 
 /**
- * Resolves the initial experimental DTS configuration
- * into a consistent
- * {@linkcode NormalizedExperimentalDtsConfig | experimentalDts config object}.
- *
- * This function handles different types of
- * {@linkcode NormalizedExperimentalDtsConfig | experimentalDts} inputs:
- * - If {@linkcode experimentalDts} is a `boolean`:
- *   - if `true`, it returns a default object with an empty entry (`{ entry: {} }`).
- *   - if `false`, it returns `undefined`.
- * - If {@linkcode experimentalDts} is a `string`, it treats the string as a glob pattern, resolving it to entry paths and returning an object with the `entry` property.
- * - If {@linkcode experimentalDts} is already an object ({@linkcode NormalizedExperimentalDtsConfig}), it resolves {@linkcode NormalizedExperimentalDtsConfig.entry | the entry paths} (if necessary) and returns the object with the updated entries.
- *
- * The function focuses specifically on normalizing the **initial**
- * {@linkcode NormalizedExperimentalDtsConfig | experimentalDts configuration}.
- *
- * @param experimentalDts - The {@linkcode NormalizedExperimentalDtsConfig | experimentalDts} value, which can be a `boolean`, `string`, `object`, or `undefined`.
- * @returns A {@linkcode Promise | promise} that resolves to a normalized {@linkcode NormalizedExperimentalDtsConfig | experimentalDts config object}, or `undefined` if the input was `false` or `undefined`.
+ * Resolves the initial experimental DTS configuration into a consistent
+ * {@link NormalizedExperimentalDtsConfig} object.
  *
  * @internal
  */
 export const resolveInitialExperimentalDtsConfig = async (
   experimentalDts: Options['experimentalDts'],
-): Promise<NormalizedOptions['experimentalDts']> => {
+): Promise<NormalizedExperimentalDtsConfig | undefined> => {
   if (experimentalDts == null) {
     return
   }
@@ -428,6 +406,8 @@ export const resolveInitialExperimentalDtsConfig = async (
     return experimentalDts ? { entry: {} } : undefined
 
   if (typeof experimentalDts === 'string') {
+    // Treats the string as a glob pattern, resolving it to entry paths and
+    // returning an object with the `entry` property.
     return {
       entry: convertArrayEntriesToObjectEntries(await glob(experimentalDts)),
     }

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -1,9 +1,16 @@
 import fs from 'node:fs'
 import path from 'node:path'
 import resolveFrom from 'resolve-from'
+import type { InputOption } from 'rollup'
 import strip from 'strip-json-comments'
 import { glob } from 'tinyglobby'
-import type { Entry, Format } from './options'
+import type {
+  Entry,
+  Format,
+  NormalizedExperimentalDtsConfig,
+  NormalizedOptions,
+  Options,
+} from './options'
 
 export type MaybePromise<T> = T | Promise<T>
 
@@ -241,4 +248,197 @@ export function trimDtsExtension(fileName: string) {
 export function writeFileSync(filePath: string, content: string) {
   fs.mkdirSync(path.dirname(filePath), { recursive: true })
   fs.writeFileSync(filePath, content)
+}
+
+/**
+ * Replaces TypeScript declaration file
+ * extensions (`.d.ts`, `.d.mts`, `.d.cts`)
+ * with their corresponding JavaScript variants (`.js`, `.mjs`, `.cjs`).
+ *
+ * @param dtsFilePath - The file path to be transformed.
+ * @returns The updated file path with the JavaScript extension.
+ *
+ * @internal
+ */
+export function replaceDtsWithJsExtensions(dtsFilePath: string) {
+  return dtsFilePath.replace(
+    /\.d\.(ts|mts|cts)$/,
+    (_, fileExtension: string) => {
+      switch (fileExtension) {
+        case 'ts':
+          return '.js'
+        case 'mts':
+          return '.mjs'
+        case 'cts':
+          return '.cjs'
+        default:
+          return ''
+      }
+    },
+  )
+}
+
+/**
+ * Converts an array of {@linkcode NormalizedOptions.entry | entry paths}
+ * into an object where the keys represent the output
+ * file names (without extensions) and the values
+ * represent the corresponding input file paths.
+ *
+ * @param arrayOfEntries - An array of file path entries as strings.
+ * @returns An object where the keys are the output file name and the values are the input file name.
+ *
+ * @example
+ * <caption>### Convert an array of entry points to an entry point object</caption>
+ *
+ * ```ts
+ * import { defineConfig } from 'tsup'
+ *
+ * export default defineConfig({
+ *   entry: ['src/index.ts', 'src/types.ts'], // Becomes `{ index: 'src/index.ts', types: 'src/types.ts' }`
+ *   format: ['esm', 'cjs'],
+ * })
+ * ```
+ *
+ * @internal
+ */
+const convertArrayEntriesToObjectEntries = (arrayOfEntries: string[]) => {
+  const objectEntries = Object.fromEntries(
+    arrayOfEntries.map(
+      (entry) =>
+        [
+          path.posix.join(
+            ...entry
+              .split(path.posix.sep)
+              .slice(1, -1)
+              .concat(path.parse(entry).name),
+          ),
+          entry,
+        ] as const,
+    ),
+  )
+
+  return objectEntries
+}
+
+/**
+ * Resolves and standardizes entry paths into an object format.
+ * If the provided entry is a string or an array of strings, it resolves
+ * any potential glob patterns using
+ * {@linkcode glob | tiny-glob's glob function}
+ * and converts the result into an entry object.
+ * If the input is already an object, it is returned as-is.
+ *
+ * @param entryPaths - The entry paths to resolve. Can be a string, an array of strings, or an object.
+ * @returns A {@linkcode Promise} that resolves to the standardized entry paths in object format.
+ *
+ * @example
+ *
+ * ```ts
+ * import { defineConfig } from 'tsup'
+ *
+ * export default defineConfig({
+ *   entry: { index: 'src/index.ts' },
+ *   format: ['esm', 'cjs'],
+ *   experimentalDts: { entry: 'src/**\/*.ts' },
+ *   // experimentalDts: { entry: 'src/**\/*.ts' }
+ *   // becomes experimentalDts: { entry: { index: 'src/index.ts', types: 'src/types.ts } }
+ * })
+ * ```
+ *
+ * @internal
+ */
+const resolveEntryPaths = async (entryPaths: InputOption) => {
+  const resolvedEntryPaths =
+    typeof entryPaths === 'string' || Array.isArray(entryPaths)
+      ? convertArrayEntriesToObjectEntries(await glob(entryPaths))
+      : entryPaths
+
+  return resolvedEntryPaths
+}
+
+/**
+ * Resolves the
+ * {@linkcode NormalizedExperimentalDtsConfig | experimental DTS config}
+ * by resolving entry paths and merging the provided
+ * TypeScript configuration options.
+ *
+ * @param options - The options containing entry points and experimental DTS configuration.
+ * @param tsconfig - The loaded TypeScript configuration data.
+ * @returns A {@linkcode Promise | promise} that resolves to the normalized experimental DTS configuration, or `undefined` if no entry or experimental DTS option is provided.
+ *
+ * @internal
+ */
+export const resolveExperimentalDtsConfig = async (
+  options: NormalizedOptions,
+  tsconfig: any,
+): Promise<NormalizedOptions['experimentalDts']> => {
+  const resolvedEntryPaths = await resolveEntryPaths(
+    options.experimentalDts?.entry || options.entry,
+  )
+
+  // Fallback to `options.entry` if we end up with an empty object.
+  const experimentalDtsObjectEntry =
+    Object.keys(resolvedEntryPaths).length === 0
+      ? Array.isArray(options.entry)
+        ? convertArrayEntriesToObjectEntries(options.entry)
+        : options.entry
+      : resolvedEntryPaths
+
+  const normalizedExperimentalDtsConfig: NormalizedExperimentalDtsConfig = {
+    compilerOptions: {
+      ...(tsconfig.data.compilerOptions || {}),
+      ...(options.experimentalDts?.compilerOptions || {}),
+    },
+
+    entry: experimentalDtsObjectEntry,
+  }
+
+  return normalizedExperimentalDtsConfig
+}
+
+/**
+ * Resolves the initial experimental DTS configuration
+ * into a consistent
+ * {@linkcode NormalizedExperimentalDtsConfig | experimentalDts config object}.
+ *
+ * This function handles different types of
+ * {@linkcode NormalizedExperimentalDtsConfig | experimentalDts} inputs:
+ * - If {@linkcode experimentalDts} is a `boolean`:
+ *   - if `true`, it returns a default object with an empty entry (`{ entry: {} }`).
+ *   - if `false`, it returns `undefined`.
+ * - If {@linkcode experimentalDts} is a `string`, it treats the string as a glob pattern, resolving it to entry paths and returning an object with the `entry` property.
+ * - If {@linkcode experimentalDts} is already an object ({@linkcode NormalizedExperimentalDtsConfig}), it resolves {@linkcode NormalizedExperimentalDtsConfig.entry | the entry paths} (if necessary) and returns the object with the updated entries.
+ *
+ * The function focuses specifically on normalizing the **initial**
+ * {@linkcode NormalizedExperimentalDtsConfig | experimentalDts configuration}.
+ *
+ * @param experimentalDts - The {@linkcode NormalizedExperimentalDtsConfig | experimentalDts} value, which can be a `boolean`, `string`, `object`, or `undefined`.
+ * @returns A {@linkcode Promise | promise} that resolves to a normalized {@linkcode NormalizedExperimentalDtsConfig | experimentalDts config object}, or `undefined` if the input was `false` or `undefined`.
+ *
+ * @internal
+ */
+export const resolveInitialExperimentalDtsConfig = async (
+  experimentalDts: Options['experimentalDts'],
+): Promise<NormalizedOptions['experimentalDts']> => {
+  if (experimentalDts == null) {
+    return
+  }
+
+  if (typeof experimentalDts === 'boolean')
+    return experimentalDts ? { entry: {} } : undefined
+
+  if (typeof experimentalDts === 'string') {
+    return {
+      entry: convertArrayEntriesToObjectEntries(await glob(experimentalDts)),
+    }
+  }
+
+  return {
+    ...experimentalDts,
+
+    entry:
+      experimentalDts?.entry == null
+        ? {}
+        : await resolveEntryPaths(experimentalDts.entry),
+  }
 }

--- a/test/dts.test.ts
+++ b/test/dts.test.ts
@@ -385,6 +385,13 @@ test('.d.ts files should be cleaned when --clean and --experimental-dts are prov
   const filesFoo = {
     'package.json': `{ "name": "tsup-playground", "private": true }`,
     'foo.ts': `export const foo = 1`,
+    'tsconfig.json': JSON.stringify(
+      {
+        compilerOptions: { skipLibCheck: true },
+      },
+      null,
+      2,
+    ),
   }
 
   const filesFooBar = {

--- a/test/dts.test.ts
+++ b/test/dts.test.ts
@@ -258,29 +258,29 @@ test('should emit declaration files with experimentalDts', async () => {
         export function sharedFunction<T>(value: T): T | null {
           return value || null
         }
-        
+
         type sharedType = {
           shared: boolean
         }
-        
+
         export type { sharedType }
     `,
     'src/server.ts': `
         export * from './shared'
 
         /**
-         * Comment for server render function 
+         * Comment for server render function
          */
         export function render(options: ServerRenderOptions): string {
           return JSON.stringify(options)
         }
-        
+
         export interface ServerRenderOptions {
           /**
            * Comment for ServerRenderOptions.stream
-           * 
+           *
            * @public
-           * 
+           *
            * @my_custom_tag
            */
           stream: boolean
@@ -298,7 +298,7 @@ test('should emit declaration files with experimentalDts', async () => {
         import * as ServerThirdPartyNamespace from 'react-dom';
         export { ServerThirdPartyNamespace }
 
-        // Export a third party module 
+        // Export a third party module
         export * from 'react-dom/server';
 
     `,
@@ -308,7 +308,7 @@ test('should emit declaration files with experimentalDts', async () => {
         export function render(options: ClientRenderOptions): string {
           return JSON.stringify(options)
         }
-        
+
         export interface ClientRenderOptions {
           document: boolean
         }

--- a/test/experimental-dts.test.ts
+++ b/test/experimental-dts.test.ts
@@ -1,0 +1,606 @@
+import { test } from 'vitest'
+import type { Options } from '../src/index.js'
+import { getTestName, run } from './utils.js'
+
+test.for([
+  { moduleResolution: 'NodeNext', moduleKind: 'NodeNext' },
+  { moduleResolution: 'Node16', moduleKind: 'Node16' },
+  { moduleResolution: 'Bundler', moduleKind: 'ESNext' },
+  { moduleResolution: 'Bundler', moduleKind: 'Preserve' },
+  { moduleResolution: 'Node10', moduleKind: 'ESNext' },
+  { moduleResolution: 'Node10', moduleKind: 'CommonJS' },
+  { moduleResolution: 'Node', moduleKind: 'ESNext' },
+  { moduleResolution: 'Node', moduleKind: 'CommonJS' },
+] as const)(
+  "experimentalDts works with TypeScript's $moduleResolution module resolution and module set to $moduleKind",
+  async ({ moduleResolution, moduleKind }, { expect, task }) => {
+    const { getFileContent, outFiles } = await run(
+      getTestName(),
+      {
+        'src/types.ts': `export type Person = { name: string }`,
+        'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types.js'`,
+        'tsup.config.ts': `export default ${JSON.stringify(
+          {
+            name: task.name,
+            entry: { index: 'src/index.ts' },
+            format: ['esm', 'cjs'],
+            experimentalDts: true,
+          } satisfies Options,
+          null,
+          2,
+        )}`,
+        'package.json': JSON.stringify(
+          {
+            name: 'testing-experimental-dts',
+            description: task.name,
+            type: 'module',
+          },
+          null,
+          2,
+        ),
+        'tsconfig.json': JSON.stringify(
+          {
+            compilerOptions: {
+              module: moduleKind,
+              moduleResolution,
+              outDir: './dist',
+              rootDir: './src',
+              skipLibCheck: true,
+              strict: true,
+            },
+            include: ['src'],
+          },
+          null,
+          2,
+        ),
+      },
+      {
+        entry: [],
+      },
+    )
+
+    expect(outFiles).toStrictEqual([
+      '_tsup-dts-rollup.d.cts',
+      '_tsup-dts-rollup.d.ts',
+      'index.cjs',
+      'index.d.cts',
+      'index.d.ts',
+      'index.js',
+    ])
+
+    const indexDtsContent = [
+      `export { foo } from './_tsup-dts-rollup.js';`,
+      `export { Person } from './_tsup-dts-rollup.js';\n`,
+    ].join('\n')
+
+    expect(await getFileContent('dist/index.d.ts')).toStrictEqual(
+      indexDtsContent,
+    )
+
+    expect(await getFileContent('dist/index.d.cts')).toStrictEqual(
+      indexDtsContent.replaceAll(
+        `'./_tsup-dts-rollup.js'`,
+        `'./_tsup-dts-rollup.cjs'`,
+      ),
+    )
+  },
+)
+
+test('experimentalDts works when `entry` is set to an array', async ({
+  expect,
+  task,
+}) => {
+  const { getFileContent, outFiles } = await run(
+    getTestName(),
+    {
+      'src/types.ts': `export type Person = { name: string }`,
+      'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types.js'`,
+      'tsup.config.ts': `export default ${JSON.stringify(
+        {
+          name: task.name,
+          entry: ['src/index.ts'],
+          format: ['esm', 'cjs'],
+          experimentalDts: true,
+        } satisfies Options,
+        null,
+        2,
+      )}`,
+      'package.json': JSON.stringify(
+        {
+          name: 'testing-experimental-dts-entry-array',
+          description: task.name,
+          type: 'module',
+        },
+        null,
+        2,
+      ),
+      'tsconfig.json': JSON.stringify(
+        {
+          compilerOptions: {
+            outDir: './dist',
+            rootDir: './src',
+            skipLibCheck: true,
+            strict: true,
+          },
+          include: ['src'],
+        },
+        null,
+        2,
+      ),
+    },
+    {
+      entry: [],
+    },
+  )
+
+  expect(outFiles).toStrictEqual([
+    '_tsup-dts-rollup.d.cts',
+    '_tsup-dts-rollup.d.ts',
+    'index.cjs',
+    'index.d.cts',
+    'index.d.ts',
+    'index.js',
+  ])
+
+  const indexDtsContent = [
+    `export { foo } from './_tsup-dts-rollup.js';`,
+    `export { Person } from './_tsup-dts-rollup.js';\n`,
+  ].join('\n')
+
+  expect(await getFileContent('dist/index.d.ts')).toStrictEqual(indexDtsContent)
+
+  expect(await getFileContent('dist/index.d.cts')).toStrictEqual(
+    indexDtsContent.replaceAll(
+      `'./_tsup-dts-rollup.js'`,
+      `'./_tsup-dts-rollup.cjs'`,
+    ),
+  )
+})
+
+test('experimentalDts works when `entry` is set to an array of globs', async ({
+  expect,
+  task,
+}) => {
+  const { getFileContent, outFiles } = await run(
+    getTestName(),
+    {
+      'src/types.ts': `export type Person = { name: string }`,
+      'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types.js'`,
+      'tsup.config.ts': `export default ${JSON.stringify(
+        {
+          name: task.name,
+          entry: ['src/**/*.ts'],
+          format: ['esm', 'cjs'],
+          experimentalDts: true,
+        } satisfies Options,
+        null,
+        2,
+      )}`,
+      'package.json': JSON.stringify(
+        {
+          name: 'entry-array-of-globs',
+          description: task.name,
+          type: 'module',
+        },
+        null,
+        2,
+      ),
+      'tsconfig.json': JSON.stringify(
+        {
+          compilerOptions: {
+            outDir: './dist',
+            rootDir: './src',
+            skipLibCheck: true,
+            strict: true,
+          },
+          include: ['src'],
+        },
+        null,
+        2,
+      ),
+    },
+    {
+      entry: [],
+    },
+  )
+
+  expect(outFiles).toStrictEqual([
+    '_tsup-dts-rollup.d.cts',
+    '_tsup-dts-rollup.d.ts',
+    'index.cjs',
+    'index.d.cts',
+    'index.d.ts',
+    'index.js',
+    'types.cjs',
+    'types.d.cts',
+    'types.d.ts',
+    'types.js',
+  ])
+
+  const indexDtsContent = [
+    `export { foo } from './_tsup-dts-rollup.js';`,
+    `export { Person } from './_tsup-dts-rollup.js';\n`,
+  ].join('\n')
+
+  const typesDtsContent = `export { Person_alias_1 as Person } from './_tsup-dts-rollup.js';\n`
+
+  expect(await getFileContent('dist/index.d.ts')).toStrictEqual(indexDtsContent)
+
+  expect(await getFileContent('dist/index.d.cts')).toStrictEqual(
+    indexDtsContent.replaceAll(
+      `'./_tsup-dts-rollup.js'`,
+      `'./_tsup-dts-rollup.cjs'`,
+    ),
+  )
+
+  expect(await getFileContent('dist/types.d.ts')).toStrictEqual(typesDtsContent)
+
+  expect(await getFileContent('dist/types.d.cts')).toStrictEqual(
+    typesDtsContent.replaceAll(
+      `'./_tsup-dts-rollup.js'`,
+      `'./_tsup-dts-rollup.cjs'`,
+    ),
+  )
+})
+
+test('experimentalDts.entry can work independent from `options.entry`', async ({
+  expect,
+  task,
+}) => {
+  const { getFileContent, outFiles } = await run(
+    getTestName(),
+    {
+      'src/types.ts': `export type Person = { name: string }`,
+      'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types.js'`,
+      'tsup.config.ts': `export default ${JSON.stringify(
+        {
+          name: task.name,
+          entry: ['src/**/*.ts'],
+          format: ['esm', 'cjs'],
+          experimentalDts: { entry: { index: 'src/index.ts' } },
+        } satisfies Options,
+        null,
+        2,
+      )}`,
+      'package.json': JSON.stringify(
+        {
+          name: 'testing-experimental-dts-entry-can-work-independent',
+          description: task.name,
+          type: 'module',
+        },
+        null,
+        2,
+      ),
+      'tsconfig.json': JSON.stringify(
+        {
+          compilerOptions: {
+            outDir: './dist',
+            rootDir: './src',
+            skipLibCheck: true,
+            strict: true,
+          },
+          include: ['src'],
+        },
+        null,
+        2,
+      ),
+    },
+    {
+      entry: [],
+    },
+  )
+
+  expect(outFiles).toStrictEqual([
+    '_tsup-dts-rollup.d.cts',
+    '_tsup-dts-rollup.d.ts',
+    'index.cjs',
+    'index.d.cts',
+    'index.d.ts',
+    'index.js',
+    'types.cjs',
+    'types.js',
+  ])
+
+  const indexDtsContent = [
+    `export { foo } from './_tsup-dts-rollup.js';`,
+    `export { Person } from './_tsup-dts-rollup.js';\n`,
+  ].join('\n')
+
+  expect(await getFileContent('dist/index.d.ts')).toStrictEqual(indexDtsContent)
+
+  expect(await getFileContent('dist/index.d.cts')).toStrictEqual(
+    indexDtsContent.replaceAll(
+      `'./_tsup-dts-rollup.js'`,
+      `'./_tsup-dts-rollup.cjs'`,
+    ),
+  )
+})
+
+test('experimentalDts.entry can be an array of globs', async ({
+  expect,
+  task,
+}) => {
+  const { getFileContent, outFiles } = await run(
+    getTestName(),
+    {
+      'src/types.ts': `export type Person = { name: string }`,
+      'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types.js'`,
+      'tsup.config.ts': `export default ${JSON.stringify(
+        {
+          name: task.name,
+          entry: { index: 'src/index.ts' },
+          format: ['esm', 'cjs'],
+          experimentalDts: { entry: ['src/**/*.ts'] },
+        } satisfies Options,
+        null,
+        2,
+      )}`,
+      'package.json': JSON.stringify(
+        {
+          name: 'testing-experimental-dts-entry-array-of-globs',
+          description: task.name,
+          type: 'module',
+        },
+        null,
+        2,
+      ),
+      'tsconfig.json': JSON.stringify(
+        {
+          compilerOptions: {
+            outDir: './dist',
+            rootDir: './src',
+            skipLibCheck: true,
+            strict: true,
+          },
+          include: ['src'],
+        },
+        null,
+        2,
+      ),
+    },
+    {
+      entry: [],
+    },
+  )
+
+  expect(outFiles).toStrictEqual([
+    '_tsup-dts-rollup.d.cts',
+    '_tsup-dts-rollup.d.ts',
+    'index.cjs',
+    'index.d.cts',
+    'index.d.ts',
+    'index.js',
+    'types.d.cts',
+    'types.d.ts',
+  ])
+
+  const indexDtsContent = [
+    `export { foo } from './_tsup-dts-rollup.js';`,
+    `export { Person } from './_tsup-dts-rollup.js';\n`,
+  ].join('\n')
+
+  expect(await getFileContent('dist/index.d.ts')).toStrictEqual(indexDtsContent)
+
+  expect(await getFileContent('dist/index.d.cts')).toStrictEqual(
+    indexDtsContent.replaceAll(
+      `'./_tsup-dts-rollup.js'`,
+      `'./_tsup-dts-rollup.cjs'`,
+    ),
+  )
+})
+
+test('experimentalDts can be a string', async ({ expect, task }) => {
+  const { getFileContent, outFiles } = await run(
+    getTestName(),
+    {
+      'src/types.ts': `export type Person = { name: string }`,
+      'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types.js'`,
+      'tsup.config.ts': `export default ${JSON.stringify(
+        {
+          name: task.name,
+          entry: ['src/**/*.ts'],
+          format: ['esm', 'cjs'],
+          experimentalDts: 'src/index.ts',
+        } satisfies Options,
+        null,
+        2,
+      )}`,
+      'package.json': JSON.stringify(
+        {
+          name: 'testing-experimental-dts-can-be-a-string',
+          description: task.name,
+          type: 'module',
+        },
+        null,
+        2,
+      ),
+      'tsconfig.json': JSON.stringify(
+        {
+          compilerOptions: {
+            outDir: './dist',
+            rootDir: './src',
+            skipLibCheck: true,
+            strict: true,
+          },
+          include: ['src'],
+        },
+        null,
+        2,
+      ),
+    },
+    {
+      entry: [],
+    },
+  )
+
+  expect(outFiles).toStrictEqual([
+    '_tsup-dts-rollup.d.cts',
+    '_tsup-dts-rollup.d.ts',
+    'index.cjs',
+    'index.d.cts',
+    'index.d.ts',
+    'index.js',
+    'types.cjs',
+    'types.js',
+  ])
+
+  const indexDtsContent = [
+    `export { foo } from './_tsup-dts-rollup.js';`,
+    `export { Person } from './_tsup-dts-rollup.js';\n`,
+  ].join('\n')
+
+  expect(await getFileContent('dist/index.d.ts')).toStrictEqual(indexDtsContent)
+
+  expect(await getFileContent('dist/index.d.cts')).toStrictEqual(
+    indexDtsContent.replaceAll(
+      `'./_tsup-dts-rollup.js'`,
+      `'./_tsup-dts-rollup.cjs'`,
+    ),
+  )
+})
+
+test('experimentalDts can be a string of glob pattern', async ({
+  expect,
+  task,
+}) => {
+  const { getFileContent, outFiles } = await run(
+    getTestName(),
+    {
+      'src/types.ts': `export type Person = { name: string }`,
+      'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types.js'`,
+      'tsup.config.ts': `export default ${JSON.stringify(
+        {
+          name: task.name,
+          entry: { index: 'src/index.ts' },
+          format: ['esm', 'cjs'],
+          experimentalDts: 'src/**/*.ts',
+        } satisfies Options,
+        null,
+        2,
+      )}`,
+      'package.json': JSON.stringify(
+        {
+          name: 'testing-experimental-dts-can-be-a-string-of-glob-pattern',
+          description: task.name,
+          type: 'module',
+        },
+        null,
+        2,
+      ),
+      'tsconfig.json': JSON.stringify(
+        {
+          compilerOptions: {
+            outDir: './dist',
+            rootDir: './src',
+            skipLibCheck: true,
+            strict: true,
+          },
+          include: ['src'],
+        },
+        null,
+        2,
+      ),
+    },
+    {
+      entry: [],
+    },
+  )
+
+  expect(outFiles).toStrictEqual([
+    '_tsup-dts-rollup.d.cts',
+    '_tsup-dts-rollup.d.ts',
+    'index.cjs',
+    'index.d.cts',
+    'index.d.ts',
+    'index.js',
+    'types.d.cts',
+    'types.d.ts',
+  ])
+
+  const indexDtsContent = [
+    `export { foo } from './_tsup-dts-rollup.js';`,
+    `export { Person } from './_tsup-dts-rollup.js';\n`,
+  ].join('\n')
+
+  expect(await getFileContent('dist/index.d.ts')).toStrictEqual(indexDtsContent)
+
+  expect(await getFileContent('dist/index.d.cts')).toStrictEqual(
+    indexDtsContent.replaceAll(
+      `'./_tsup-dts-rollup.js'`,
+      `'./_tsup-dts-rollup.cjs'`,
+    ),
+  )
+})
+
+test('experimentalDts.entry can be a string of glob pattern', async ({
+  expect,
+  task,
+}) => {
+  const { getFileContent, outFiles } = await run(
+    getTestName(),
+    {
+      'src/types.ts': `export type Person = { name: string }`,
+      'src/index.ts': `export const foo = [1, 2, 3]\nexport type { Person } from './types.js'`,
+      'tsup.config.ts': `export default ${JSON.stringify(
+        {
+          name: task.name,
+          entry: { index: 'src/index.ts' },
+          format: ['esm', 'cjs'],
+          experimentalDts: { entry: 'src/**/*.ts' },
+        } satisfies Options,
+        null,
+        2,
+      )}`,
+      'package.json': JSON.stringify(
+        {
+          name: 'testing-experimental-dts-entry-can-be-a-string-of-glob-pattern',
+          description: task.name,
+          type: 'module',
+        },
+        null,
+        2,
+      ),
+      'tsconfig.json': JSON.stringify(
+        {
+          compilerOptions: {
+            outDir: './dist',
+            rootDir: './src',
+            skipLibCheck: true,
+            strict: true,
+          },
+          include: ['src'],
+        },
+        null,
+        2,
+      ),
+    },
+    {
+      entry: [],
+    },
+  )
+
+  expect(outFiles).toStrictEqual([
+    '_tsup-dts-rollup.d.cts',
+    '_tsup-dts-rollup.d.ts',
+    'index.cjs',
+    'index.d.cts',
+    'index.d.ts',
+    'index.js',
+    'types.d.cts',
+    'types.d.ts',
+  ])
+
+  const indexDtsContent = [
+    `export { foo } from './_tsup-dts-rollup.js';`,
+    `export { Person } from './_tsup-dts-rollup.js';\n`,
+  ].join('\n')
+
+  expect(await getFileContent('dist/index.d.ts')).toStrictEqual(indexDtsContent)
+
+  expect(await getFileContent('dist/index.d.cts')).toStrictEqual(
+    indexDtsContent.replaceAll(
+      `'./_tsup-dts-rollup.js'`,
+      `'./_tsup-dts-rollup.cjs'`,
+    ),
+  )
+})

